Enhanced cookbook recipe validation

juanpprieto opened the pull request:
**Fixes https://github.com/Shopify/hydrogen-internal/issues/248**

## Summary

Comprehensive validation system for Hydrogen cookbook recipes with detailed error reporting, TypeScript strict checking, and proper step hierarchy enforcement. Catches errors before CI with actionable messages including line numbers and actual values.

## Changes

### Validation System

**Error Reporting:**
- Line numbers for every error (e.g., `recipe.yaml:52`)
- Actual YAML values shown (e.g., `(actual value: "1")`)
- Validator names prefixed (e.g., `RecipeSchema:`, `validatePatchFiles:`)
- All errors collected and displayed together
- Consistent formatting across all cookbook commands

**Validators:**
- Schema validation with enhanced Zod error formatting
- Step number uniqueness (rejects duplicates, enforces substep hierarchy)
- Step descriptions required (non-null, non-empty)
- Bidirectional file validation (patches and ingredients)
- Documentation completeness (README.md and LLM prompts)

**Example Output:**
```
❌ Recipe 'gtm' - 3 error(s):

recipe.yaml:52      steps.0.step                  RecipeSchema: Expected string, received number (actual value: 1)
                    README.md                     validateReadmeExists: README.md not found. Run: npm run cookbook render gtm
                    llms/gtm.prompt.md            validateLlmPromptExists: LLM prompt not found. Run: npm run cookbook render gtm
```

### Step Hierarchy

Duplicate step numbers are now rejected. Use substeps for organization:

```yaml
# ✅ VALID - Proper hierarchy
steps:
  - step: "1"
    type: INFO
    name: Setup configuration
  - step: "1.1"
    name: README.md
  - step: "1.2"
    name: app/root.tsx
  - step: "2"
    name: package.json

# ❌ INVALID - Duplicate step numbers
steps:
  - step: "1"
    name: README.md
  - step: "1"
    name: app/root.tsx
```

**Rendering:**
```markdown
### Step 1: Setup configuration
#### Step 1.1: README.md
#### Step 1.2: app/root.tsx
### Step 2: package.json
```

### TypeScript Integration

- Added `typecheck` script to cookbook/package.json
- Configured tsconfig.json with strict checking and skipLibCheck
- Fixed all Recipe type mocks with proper llms field
- Fixed vitest mock types using Mock type with assertions
- Aligned TypeScript version to 5.9.2 (matches packages/)
- Zero TypeScript errors in cookbook/src/

### Error Handling

All commands now show formatted errors instead of raw Zod dumps:
- apply, render, regenerate, update, LLM generation
- Exported `printValidationErrors` and `handleZodErrorFromLoadRecipe` helpers
- Consistent error formatting across entire cookbook system

### Test Coverage

46 tests covering:
- Schema validation with actual values
- Step number uniqueness (rejects "1", "1"; allows "1", "1.1", "1.2")
- Numeric equivalence detection ("1" === "01" === "1.0")
- Substep hierarchy validation
- File reference validation (bidirectional)
- Documentation completeness
- Numeric sorting (1 < 2 < 10, not lexicographic)
- Edge cases (zero steps, large numbers, substeps)

## Breaking Changes

Recipe step field changed from `number` to `string` with regex validation `/^\d+(?:\.\d+)?$/`:

```diff
steps:
-  - step: 1
+  - step: "1"
    name: README.md
```

This enables substep notation ("1.1", "1.2") while maintaining numeric sorting.

## Testing

```bash
cd cookbook
npm install
npm test           # 46/46 passing
npm run typecheck  # Zero errors
npm run lint       # Zero warnings
npm run cookbook -- validate --recipe {name}
```

All validation, tests, typecheck, and linting integrated with turbo monorepo workflow.


### New validation/render/generate tests

## Test Environment
- Recipe: express
- Commands: render (github/shopify.dev), validate

---

## Test 1: Substep Hierarchy Rendering

**Modified recipe.yaml:**
```yaml
steps:
  - type: INFO
    step: "1"
    name: Initial configuration
  - type: PATCH
    step: "1.1"
    name: Disable customer account API
  - type: PATCH
    step: "1.2"
    name: Update README
  - type: NEW_FILE
    step: "1.3"
    name: Add environment types
  - type: INFO
    step: "2"
    name: Configure client and server rendering
  - type: PATCH
    step: "2.1"
    name: Set up client-side hydration
  - type: NEW_FILE
    step: "2.2"
    name: Add favicon
  - type: PATCH
    step: "2.3"
    name: Configure server-side rendering
```

**Command:** `npm run cookbook -- render --recipe express --format github`

**GitHub README Output (lines 40-320):**
```
## Steps

### Step 1: Initial configuration
#### Step 1.1: Disable customer account API
##### File: .graphqlrc.ts
#### Step 1.2: Update README for Express deployment
##### File: README.md
#### Step 1.3: Add environment type definitions
##### File: env.ts
### Step 2: Configure client and server rendering
#### Step 2.1: Set up client-side hydration
##### File: app/entry.client.tsx
#### Step 2.2: Add the Express template favicon
##### File: favicon.svg
#### Step 2.3: Configure server-side rendering
##### File: app/entry.server.tsx
```

**Shopify.dev Output (lines 47-322):**
```
## Step 1: Initial configuration
### Step 1.1: Disable customer account API
##### File: .graphqlrc.ts
### Step 1.2: Update README for Express deployment
##### File: README.md
### Step 1.3: Add environment type definitions
##### File: env.ts
## Step 2: Configure client and server rendering
### Step 2.1: Set up client-side hydration
##### File: app/entry.client.tsx
```

**Result:** ✅ PASS
- GitHub: h3 parent → h4 substeps → h5 files
- Shopify.dev: h2 parent → h3 substeps → h5 files
- Perfect hierarchy in both formats

---

## Test 2: Duplicate Step Number Detection

**Modified recipe.yaml (line 107):**
```yaml
steps:
  - type: INFO
    step: "1"
    name: Initial configuration
  - type: PATCH
    step: "1"              # ← Changed from "1.1" to "1" (DUPLICATE!)
    name: Disable customer account API
```

**Command:** `npm run cookbook -- validate --recipe express`

**Output:**
```
❌ Recipe 'express' - 1 error(s):

recipe.yaml:107     steps[1].step                 validateStepNames: Duplicate step number "1". Use substeps (e.g., "1.1", "1.2") to organize related changes under step 1.

❌ Recipe 'express' is invalid
```

**Result:** ✅ PASS
- Detected duplicate at exact line (107)
- Clear error message with suggestion
- Validation prevented invalid recipe

---

juanpprieto committed:
test(cookbook): add vitest testing infrastructure

Set up vitest for cookbook package validation tests following repo patterns:
- Add vitest dependency and test scripts
- Create vitest.config.ts with node environment
- Add smoke test (util.test.ts) with 2 passing tests
- Add cookbook to root workspaces for turbo integration

Tests can be run via:
- `cd cookbook && npm test`
- `npx turbo run test --filter=cookbook`

juanpprieto committed:
feat(cookbook): enforce numeric step numbers and validate duplicate names

Enforces step numbers as numbers (not strings) via Zod schema and adds
runtime validation to catch duplicate step names in grouped steps.

Changes:
- Update Zod schema to only accept numeric step numbers
- Update JSON schema to match (remove string union type)
- Add validateStepNumbering() to check for duplicate names in grouped steps
- Integrate validation into validateRecipe() flow before applyRecipe()
- Add comprehensive test coverage (7 new tests)

This catches the PR-3228 bug where recipe.yaml had step: "1" as strings
and prevents duplicate step names like:
  step: 1, name: "Update config"
  step: 1, name: "Update config"  ← now throws error

Grouped steps with distinct names are still allowed:
  step: 1, name: "README.md"
  step: 1, name: "app/entry.server.tsx"  ← valid

Performance: Recipe is loaded twice (once for validation, once in applyRecipe).
This duplication is acceptable as the overhead is negligible (~5ms) and keeps
the implementation simple. Can be optimized later if needed.

Tests: 9 passing (3 Zod validation tests, 4 duplicate name tests, 2 smoke tests)

juanpprieto committed:
refactor(cookbook): split validators and enforce step descriptions

Refactors validation into focused, single-responsibility validators:
- validateStepNames() - checks for duplicate names in grouped steps
- validateStepDescriptions() - enforces descriptions are not null/empty

This is a breaking change that will fail validation for 55 steps across 7
recipes (b2b, metaobjects, legacy-customer-account-flow, gtm, infinite-scroll,
combined-listings, third-party-api) that currently have description: null.

These will be fixed in a separate PR per PR-3228 discussion.

Tests: 11 passing (4 step name tests, 2 description tests, 5 other tests)

juanpprieto committed:
feat(cookbook): add bidirectional file reference validation

Adds validators to ensure patch and ingredient files match between recipe.yaml
and filesystem in both directions - no broken references, no orphaned files.

Validators added:
- validatePatchFiles() - checks patches referenced in yaml exist on disk,
  and all patch files on disk are referenced in yaml
- validateIngredientFiles() - checks ingredients referenced in yaml exist on
  disk, and all ingredient files on disk are referenced in yaml

This catches:
- Missing patch/ingredient files (broken references)
- Orphaned patch/ingredient files (not referenced in recipe)
- Early detection before recipe application

Performance: O(n) single-pass with Set for O(1) lookups, early returns when
directories don't exist.

Tests: 17 passing (12 validation tests, 3 recipe tests, 2 util tests)
  - 3 patch validation tests
  - 3 ingredient validation tests
  - 6 existing validation tests

juanpprieto committed:
feat(cookbook): validate README and LLM prompt existence

Adds validators to ensure rendered outputs exist:
- validateReadmeExists() - checks README.md exists at recipes/{name}/README.md
- validateLlmPromptExists() - checks LLM prompt exists at llms/{name}.prompt.md

Uses real integration tests against actual filesystem (gtm recipe) instead of
mocked tests. This provides more confidence that validation works correctly.

Changes:
- Add LLMS_PATH and RENDER_FILENAME_GITHUB to imports
- Add two focused validators with helpful error messages
- Integrate into validateRecipe() flow
- Remove global fs mock, use targeted spyOn mocks only where needed
- Add 4 integration tests (2 per validator)

Tests: 21 passing (16 validation tests, 3 recipe tests, 2 util tests)

juanpprieto committed:
fix(cookbook): use numeric step fallback in generate

Fixes TypeScript error caused by string step number fallback after enforcing
numeric-only step numbers in recipe schema.

Changed: `step: existingStep?.step ?? \`${i}\``
To: `step: existingStep?.step ?? i`

This ensures generated steps always use numbers, matching the Zod schema
requirement from commit c937534d5.

juanpprieto committed:
feat(cookbook): enhance validation error reporting with actual values

Improves validation error messages by showing actual YAML values alongside
expected types, making it immediately clear what needs to be fixed.

Changes:
- Add getYamlValue() to extract actual values from YAML nodes
- Include actual values in Zod schema error messages
  Example: "Expected number, received string (actual value: "1")"
- Add validator names to all error output for clarity
- Refactor error collection to run all validators even with schema errors
  (README/LLM prompt validators now run regardless of recipe schema validity)
- Remove unused functions: convertZodErrorToValidationErrors, enrichErrorsWithLineNumbers
- Add comprehensive unit tests for formatValidationError
- Add integration test validating actual values in error messages
- Mock filesystem in validateReadmeExists and validateLlmPromptExists tests

Test coverage: 26/26 tests passing

juanpprieto committed:
test(cookbook): mock filesystem in actual value validation test

Fixes CI failure where test relied on actual filesystem paths that differ
between local and CI environments. Now uses mocked YAML content instead.

juanpprieto committed:
test(cookbook): mock filesystem in getYamlLineNumber tests

Ensures tests work in both local and CI environments by mocking
fs.readFileSync instead of relying on actual recipe files.

juanpprieto committed:
feat(cookbook)!: migrate step field from number to string with regex validation

BREAKING CHANGE: Recipe step field now requires string format instead of number.
All recipe.yaml files must be updated: step: 1 → step: "1"

Changes:
- Update Zod schema: z.number() → z.string().regex(/^\d+(?:\.\d+)?$/)
- Update JSON schema: type "number" → type "string" with pattern validation
- Add comprehensive test coverage (39 tests passing)
- Fix compareSteps bug: isSubstep(a) → isSubstep(step)
- Remove unnecessary String() conversions in validation
- Export compareSteps for testability

Test improvements:
- Add generate.test.ts with step ordering tests
- Update all test expectations for string steps
- Add edge case validation (invalid formats, substeps)
- Remove unused leading zeros test
- Document grouping feature and testing strategy

rbshop commented on cookbook/src/lib/generate.test.ts:
```diff
@@ -0,0 +1,124 @@
+
+  // Grouping feature: Multiple steps can share the same step number to organize
+  // related changes together (e.g., modifying several files as part of "Step 1").
+  // This test ensures sort maintains relative order for grouped steps.
```
I'm pretty sure this is wrong

rbshop commented on cookbook/src/lib/generate.ts:
```diff
@@ -183,7 +183,7 @@ async function generateSteps(params: {
       const step: Step = {
         type: 'PATCH',
-        step: existingStep?.step ?? `${i}`,
+        step: existingStep?.step ?? i,
```
Now that steps are strings this should be reverted

juanpprieto committed:
fix(cookbook): address PR feedback and add TypeScript strict checking

Fixes issues identified in PR review and adds comprehensive TypeScript checking.

Changes:
- Fix type mismatch in generate.ts: use String(i) for step field fallback
- Add explanatory comment for empty catch block in validate.ts
- Fix test to verify actual relative order preservation in compareSteps
- Add 3 edge case tests for compareSteps (leading zeros, mixed types, decimals)
- Fix all Recipe test mocks: add missing llms field (12+ instances)
- Fix TypeScript mock types: use Mock type with assertions
- Add typecheck script and configure tsconfig with skipLibCheck
- Align TypeScript version to 5.9.2 (matches packages/)

Test coverage: 42/42 tests passing
TypeScript: Zero errors in src/

juanpprieto committed:
test(cookbook): add comprehensive grouping test coverage and cleanup test names

Adds 9 new tests covering edge cases and cleans up test descriptions for clarity.

New tests:
- Multiple separate groups with stable sort
- Interleaved groups maintaining insertion order
- Interleaved main steps and substeps (validates numeric sorting)
- Multiple substeps with same step number
- Explicit "1.0" vs implicit "1" equivalence
- Zero step numbers and zero with substeps
- Very large step numbers (9999, 1000)
- Multiple files in same step number (real-world scenario)

Test name improvements:
- Remove verbose comments and investigation tags
- Simplify descriptions for contributor clarity
- Focus on what behavior is being validated

Coverage: 51/51 tests passing

juanpprieto committed:
feat(cookbook): add formatted error handling to all loadRecipe calls

Ensures consistent error formatting across all commands that load recipes.

Changes:
- Export printValidationErrors for reuse across commands
- Add handleZodErrorFromLoadRecipe helper to format Zod schema errors
- Wrap loadRecipe calls in try-catch in apply.ts, render.ts, llms.ts, update.ts
- Display errors with line numbers and actual values instead of raw Zod output

Before:
  ZodError: [{ code: 'invalid_type', expected: 'string', received: 'number', path: ['steps', 0, 'step'], ... }]

After:
  recipe.yaml:34  steps.0.step  RecipeSchema: Expected string, received number (actual value: 1)

All commands now have consistent, actionable error messages.

rbshop committed:
Merge branch 'main' into enhance-cookbook-validate

rbshop committed:
Merge branch 'main' into enhance-cookbook-validate

rbshop commented on .changeset/enforce-calver-ci.js:
```diff
@@ -21,7 +21,7 @@ const {
   getPackagePath,
   readPackage,
-  writePackage
+  writePackage,
```
Not sure what happened here, but this file's addition appears to be just from formatting changes

rbshop commented on cookbook/src/lib/generate.test.ts:
```diff
@@ -0,0 +1,311 @@
+      {type: 'PATCH', step: '1', name: 'File A', diffs: []},
+      {type: 'PATCH', step: '1', name: 'File B', diffs: []},
+      {type: 'PATCH', step: '2', name: 'File D', diffs: []},
+    ];
```
This might have been the intention at some point, but since this breaks rendering we should either prevent duplicate step numbers as part of the validation, or fix the rendering. I'd say it makes more sense to prevent via validation, since we have substep support already

juanpprieto committed:
feat(cookbook): reject duplicate step numbers, enforce substep hierarchy

Addresses PR feedback to prevent confusing duplicate step number headings.

Changes:
- Reject duplicate step numbers (e.g., two steps both labeled '1')
- Detect numeric equivalence ('1' === '01' === '1.0')
- Provide helpful error messages suggesting substeps
- Add 4 new validation tests for duplicate detection
- Remove 5 grouping tests that validated incorrect behavior
- Simplify generate.test.ts to 12 focused tests
- Add lint script to cookbook/package.json

Validation now enforces proper hierarchy:
  ✅ VALID:   step: '1' → step: '1.1' → step: '1.2' → step: '2'
  ❌ INVALID: step: '1' → step: '1' (use substeps instead)

Rendering output:
  ### Step 1: Setup
  #### Step 1.1: README.md
  #### Step 1.2: app/root.tsx

Test coverage: 46/46 tests passing
